src/System.Management.Automation/engine/CoreAdapter.cs

```diff
@@ -1394,32 +1394,37 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
             // be turned into an array.
             // We also skip the optimization if the number of arguments and parameters is different
             // so we let the loop deal with possible optional parameters.
-            if ((methods.Length == 1) &&
-                (!methods[0].hasVarArgs) &&
-                (!methods[0].isGeneric) &&
-                (methods[0].method == null || !(methods[0].method.DeclaringType.IsGenericTypeDefinition)) &&
+            if (methods.Length == 1
+                && !methods[0].hasVarArgs
                 // generic methods need to be double checked in a loop below - generic methods can be rejected if type inference fails
-                (methods[0].parameters.Length == arguments.Length))
+                && !methods[0].isGeneric
+                && (methods[0].method is null || !(methods[0].method.DeclaringType.IsGenericTypeDefinition))
+                && methods[0].parameters.Length == arguments.Length)
             {
                 return methods[0];
             }
 
             Type[] argumentTypes = arguments.Select(EffectiveArgumentType).ToArray();
+            Type[] genericParameters = invocationConstraints?.GenericTypeParameters ?? Array.Empty<Type>();
```

Comment:
The `?? Array.Empty<Type>();` part is not needed, as your changes to `PSMethodInvocationConstraints` constructor already makes sure it's always initialized with an empty array in case it's null.
But I don't know, maybe we should keep it here and remove it from the constructor.

Reply:
> I think it's possible for invocationConstraints itself here to be null

Yeah, I agree. Then let's remove it from the constructor.

Reply:
Addressed.